README.adoc

```diff
@@ -0,0 +1,161 @@
+
+==== Deploy on your own Kubernetes
+
+In case you prefer to host your own services. You need to provide a Kubernetes cluster. This cluster could be local or cloud-based and must fulfill the following requirements:
+
+* A Kubernetes cluster, with 1.13 as the minimum required version
+* An ingress controller with a Layer7 Load Balancer ability
+* A working knowledge of Kubernetes to be able to manage your cluster and to troubleshoot, should any issue specific to your Kubernetes cluster arise.
+
+
+== Build Docker images
+
+Before beginning the deployment, the various microservices composing the chat application must be packaged each into their own docker image, and published to your docker registry.
+
+=== Building the Friends & Presence services
+* https://github.com/cloudstateio/samples-js-chat/blob/master/friends/README.adoc#building-the-friends-service[Building the Friends service]
+
+* https://github.com/cloudstateio/samples-js-chat/blob/master/presence/README.adoc#building-the-presence-service[Building the Presence service]
+
+=== Building the Chat-UI service
+
+The code below assumes that your docker registry is named `mydockerregistry`. Please replace it with your DockerID (which you have registed at https://hub.docker.com/ as mentioned in <<infrastructure>>
+
+
+[source,shell]
+----
+cd ./chat-ui
+
+nvm install
+nvm use
+npm install
+./protogen.sh
+npm run prestart
+npm run-script build
+
+# let's assume your DockerID in https://hub.docker.com/ is `mydockerregistry`
+DOCKER_PUBLISH_TO=mydockerregistry
+
+# build docker image
+docker build . -t $DOCKER_PUBLISH_TO/samples-js-chat-ui:latest
+
+# authenticate with your Docker registry
+docker login
+
+# push the docker image to your registry
+docker push $DOCKER_PUBLISH_TO/samples-js-chat-ui:latest
+----
+
+=== Testing the Chat-UI service
+
+In order to make the deployment of the Chat UI consistent with all the other services composing the Chat application. The Node.js server running the UI code is wrapped in a Cloudstate stateless service.
+
+As such, the UI must be accessed via a Cloudstate proxy. The procedure is similar to that of https://github.com/cloudstateio/samples-js-chat/blob/master/friends/README.adoc#testing-friends-service[Testing Friends service]. Which means:
+
+* Start a Cloudstate proxy (docker image prebuilt for us by Cloudstate dev team)
+* Start the docker image of the Chat UI we have just built
+* Access the UI via the Cloudstate proxy, which expose the Chat-UI service on `http://localhost:9000/pages/chat.html`
+
+[source,shell]
+----
+# terminal 1: run the cloudstate proxy docker image
+# exposing the CS proxy port 9000 to the same 9000 port on the host machine
+# NOTE: ignore the connection errors b/c the `user-function` is not yet started
+docker run -it --rm --name cloudstate-proxy -p 9000:9000 \
+  cloudstateio/cloudstate-proxy-dev-mode \
```

Comment:
`cloudstateio/cloudstate-proxy-dev-mode:latest` is the build from master, no?

Reply:
I don't know well this part, if there is a recommended Cloudstate proxy other than `cloudstateio/cloudstate-proxy-dev-mode:latest` can you please suggest?

Reply:
Yes, `latest` is the build from master. It could be better to have `latest` point to the latest release instead, and use a different tag for the builds from master. For now, how about using the latest release: `cloudstateio/cloudstate-proxy-dev-mode:0.5.1`

Reply:
By changing to `cloudstateio/cloudstate-proxy-dev-mode:0.5.1`, the documentation may get out-of-date in a near future upon a new release of the Cloudstate proxy. Can you confirm which approach is more preferable?
1. `latest`: doc always up-to-date but may run into instabilities b/c Cloudstate proxy build on `master` is not of release quality
2. `0.5.1` is guarantee stable, but we must remember to update all docs. The difficulty here is that the maintainers of the samples code are likely not the same as those maintaining Cloudstate.

Reply:
The sample depends on a particular version of cloudstate node-support (currently `0.0.3`). If changes are made to the proxy, then the sample may not work. So I think it's better to have a stable version for the proxy which is known to work. Since this is described as a maintenance issue, we should focus there — ideally we have automated or easy ways to update all the samples after releases.